collectors/htlcs_collector.go

```diff
@@ -0,0 +1,209 @@
+package collectors
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/lightninglabs/lndclient"
+	"github.com/lightningnetwork/lnd/htlcswitch"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+// htlcMonitor contains the elements required to monitor our htlc forwarding
+// stream. Since we collect metrics from a stream, rather than a set of custom
+// polled metrics, we just use the built-in prometheus types to monitor our
+// forwards, and do not implement the collector interface.
+type htlcMonitor struct {
+	// router provides us with access to lnd's router rpc.
+	router lndclient.RouterClient
+
+	// successCounter is a counter which tracks the number of successful
+	// forwards through the node.
+	successCounter prometheus.Counter
+
+	// failureCounter is a counter which tracks the number of failed
+	// forwards through the node.
+	failureCounter prometheus.Counter
+
+	// activeForwards holds a map of our currently active htlcs related
+	// to forwards.
+	activeForwards map[htlcswitch.HtlcKey]struct{}
+
+	// quit is closed to signal that we need to shutdown.
+	quit chan struct{}
+
+	wg sync.WaitGroup
+
+	// errChan is a channel that we send any errors that we encounter into.
+	// This channel should be buffered so that it does not block sends.
+	errChan chan<- error
+}
+
+func newHtlcMonitor(router lndclient.RouterClient,
+	errChan chan error) *htlcMonitor {
+
+	return &htlcMonitor{
+		router:         router,
+		activeForwards: make(map[htlcswitch.HtlcKey]struct{}),
+		successCounter: prometheus.NewCounter(prometheus.CounterOpts{
```

Comment:
Some useful labels for both of these (so a counter vec) include:
  * htlc amt
  * proposed htlc free
  * incoming chan 
  * outgoing chan 
  * payload version

Reply:
✅ incoming/outgoing channel labels added 
❌  Payload version not currently exposed by lnd. 

What kind of queries would amount/fee cover, snce these labels would be string values, (I think) we wouldn't be able to use them very efficiently? Perhaps better to track separately in fee/amt histograms? Also going to end up with a lot of time series if we track for each chan/amt/fee combination.

Reply:
Ah yeh since they end up as strings, they aren't as useful for manipulation. If we wanted all this data we would instead need to export them as individual time series, then use aggregation operators based on a payment hash, etc. The incoming/outgoing chan IDs are indeed the only ones that would actually useable!